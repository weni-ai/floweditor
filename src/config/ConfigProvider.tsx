import { FlowEditorConfig } from 'flowTypes';
import * as React from 'react';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
  config: FlowEditorConfig;
  children: any;
}

export interface ConfigProviderContext {
  config: FlowEditorConfig;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR =
  'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
  'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<
  ConfigProviderProps
> {
  public static childContextTypes = {
    config: fakePropType,
    assetService: fakePropType,
  };

  constructor(props: ConfigProviderProps) {
    super(props);

    if (React.Children.count(props.children) > 1) {
      throw new Error(SINGLE_CHILD_ERROR);
    } else if (!React.isValidElement(props.children)) {
      throw new Error(VALID_CHILD_ERROR);
    }
  }

  public getChildContext(): ConfigProviderContext {
    return {
      config: this.props.config,
    };
  }

  public render(): JSX.Element {
    return React.Children.only(this.props.children);
  }
}
