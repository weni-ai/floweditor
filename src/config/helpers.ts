import {
  FlowTypes,
  FlowTypeVisibility,
  Operator,
  Type,
  Types,
  VISIBILITY_ONLINE,
} from 'config/interfaces';
import { FlowEditorConfig } from 'flowTypes';
import i18n from './i18n';
import { NodeEditorSettings } from '../store/nodeEditor';

export const isOnlineFlowType = (flowType: FlowTypes) => {
  return !!VISIBILITY_ONLINE.find((type: FlowTypes) => type === flowType);
};

export const filterOperators = (
  operators: Operator[],
  config: FlowEditorConfig,
): Operator[] => {
  return filterVisibility(operators, config);
};

export const filterTypeConfigs = (
  typeConfigs: Type[],
  config: FlowEditorConfig,
  nodeSettings: NodeEditorSettings,
): Type[] => {
  return filterVisibility(typeConfigs, config, nodeSettings).map(
    (typeConfig: Type) => {
      if (typeConfig.new) {
        return {
          ...typeConfig,
          description: `${typeConfig.description} (${i18n.t('new', 'New')})`,
        };
      }

      return typeConfig;
    },
  );
};

const filterVisibility = (
  items: FlowTypeVisibility[],
  config: FlowEditorConfig,
  nodeSettings?: NodeEditorSettings,
): any[] => {
  return items.filter((item: FlowTypeVisibility) => {
    if (nodeSettings) {
      const typeItem = item as Type;
      if (typeItem.type === Types.call_brain) {
        // check if we can show the call brain action
        // if we are not a ghost node
        // and if we are in a multi action node, we can't show the call brain action
        if (
          nodeSettings.originalNode.node.actions.length >= 1 &&
          !nodeSettings.originalNode.ghost
        ) {
          console.log('filter', nodeSettings);
          return false;
        }
      }
    }

    // if we have a filter on our type, don't return it unless its present in our config
    if (item.filter) {
      if (
        !(config.filters || []).find((name: string) => name === item.filter)
      ) {
        return false;
      }
    }

    if (item.visibility === undefined) {
      return true;
    }

    return (
      item.visibility.findIndex((ft: FlowTypes) => ft === config.flowType) > -1
    );
  });
};
