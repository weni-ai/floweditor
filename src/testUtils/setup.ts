/* eslint-disable no-redeclare */
import { Console } from 'console';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Exit } from 'flowTypes';
import { RenderNode } from 'store/flowContext';

// Declare custom matcher types
declare global {
  namespace jest {
    interface Matchers<R> {
      toPointTo(renderNode: RenderNode): R;
      toHaveExitThatPointsTo(renderNode: RenderNode): R;
      toHaveInboundFrom(exit: Exit): R;
      toHaveExitWithDestination(): R;
      toHaveInboundConnections(): R;
      toHavePayload(action: string, payload: any): R;
      toHaveReduxActions(actions: string[]): R;
      toMatchCallSnapshot(snapshotName?: string): R;
    }
  }
}

// Ensure console logs are visible while running tests https://github.com/facebook/jest/issues/3853
global.console = new Console(process.stderr, process.stderr);

// Configure Enzyme adapter
configure({ adapter: new Adapter() });

// RAF shim
// tslint:disable-next-line:ban-types
(global as any).requestAnimationFrame = (callback: Function) => {
  setTimeout(callback, 0);
};
