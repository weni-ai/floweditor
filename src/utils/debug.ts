import { FlowStoreProps } from 'components/flow/Flow';
import { DebugState } from 'store/editor';

import mutate from 'immutability-helper';

/* istanbul ignore next -- @preserve */
export default class Debug {
  private props: FlowStoreProps;
  private state: DebugState;

  constructor(props: FlowStoreProps, initial: DebugState) {
    this.props = props;
    this.state = initial || { showUUIDs: false };
  }

  public showUUIDs(): DebugState {
    const updated = mutate(this.state, { $merge: { showUUIDs: true } });
    this.props.mergeEditorState({ debug: updated });
    return updated;
  }
}
