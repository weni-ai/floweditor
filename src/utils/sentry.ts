import * as Sentry from '@sentry/react';

const initSentry = (dsn: string) => {
  Sentry.init({
    dsn,
    integrations: [
      Sentry.browserTracingIntegration(),
      Sentry.httpClientIntegration({
        failedRequestStatusCodes: [[400, 499], [500, 599]],
      }),
      Sentry.replayIntegration(),
      Sentry.replayCanvasIntegration(),
    ],
    tracesSampleRate: 1.0,
    replaysSessionSampleRate: 0.1,
    replaysOnErrorSampleRate: 1.0,
  });
};

export default initSentry;
