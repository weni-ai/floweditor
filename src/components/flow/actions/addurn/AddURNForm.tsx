import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { ActionFormProps } from 'components/flow/props';
import SelectElement, {
  SelectOption,
} from 'components/form/select/SelectElement';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TypeList from 'components/nodeeditor/TypeList';
import * as React from 'react';
import {
  FormState,
  mergeForm,
  SelectOptionEntry,
  StringEntry,
} from 'store/nodeEditor';
import { shouldRequireIf, validate } from 'store/validators';

import styles from './AddURNForm.module.scss';
import { getSchemeOptions, initializeForm, stateToAction } from './helpers';
import i18n from 'config/i18n';
import { renderIssues } from '../helpers';

export interface AddURNFormState extends FormState {
  scheme: SelectOptionEntry;
  path: StringEntry;
}

export const controlLabelSpecId = 'label';

export default class AddURNForm extends React.PureComponent<
  ActionFormProps,
  AddURNFormState
> {
  constructor(props: ActionFormProps) {
    super(props);
    this.state = initializeForm(this.props.nodeSettings);
    bindCallbacks(this, {
      include: [/^handle/],
    });
  }

  public handleSave(): void {
    const valid = this.handlePathChanged(this.state.path.value, null, true);
    if (valid) {
      const newAction = stateToAction(this.props.nodeSettings, this.state);
      this.props.updateAction(newAction);
      this.props.onClose(false);
    }
  }

  public handleSchemeChanged(selected: SelectOption): boolean {
    const updates: Partial<AddURNFormState> = {
      scheme: { value: selected },
    };
    const updated = mergeForm(this.state, updates);
    this.setState(updated);
    return updated.valid;
  }

  public handlePathChanged(
    value: string,
    name: string,
    submitting = false,
  ): boolean {
    const updates: Partial<AddURNFormState> = {
      path: validate(i18n.t('forms.urn'), value, [shouldRequireIf(submitting)]),
    };

    const updated = mergeForm(this.state, updates);
    this.setState(updated);
    return updated.valid;
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.confirm'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true),
      },
    };
  }

  public render(): JSX.Element {
    const typeConfig = this.props.typeConfig;
    return (
      <Dialog
        className={styles.dialog}
        title={typeConfig.name}
        headerClass={typeConfig.type}
        buttons={this.getButtons()}
      >
        <TypeList
          __className=""
          initialType={typeConfig}
          onChange={this.props.onTypeChange}
        />

        <div className={styles.scheme_selection}>
          <div
            className={`${styles.label} u font secondary body-md color-neutral-cloudy`}
          >
            {i18n.t(
              'forms.add_urn_summary',
              'Add a new URN to reach the contact such as a phone number',
            )}
          </div>

          <SelectElement
            key={'urn_type_select'}
            name={i18n.t('forms.urn_type', 'URN Type')}
            entry={this.state.scheme}
            onChange={this.handleSchemeChanged}
            options={getSchemeOptions()}
          />
        </div>

        <TextInputElement
          name={i18n.t('forms.urn')}
          showLabel={true}
          placeholder={i18n.t('forms.enter_urn_value', 'Enter the URN value')}
          entry={this.state.path}
          onChange={this.handlePathChanged}
          autocomplete={true}
        />
        {renderIssues(this.props)}
      </Dialog>
    );
  }
}
