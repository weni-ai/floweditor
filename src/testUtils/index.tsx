// TODO: Remove use of Function
// tslint:disable:ban-types
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { FlowDefinition, FlowEditorConfig } from 'flowTypes';
import mutate, { Query } from 'immutability-helper';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import createStore from 'store/createStore';
import { RenderNodeMap } from 'store/flowContext';
import { getFlowComponents } from 'store/helpers';
import AppState, { initialState } from 'store/state';
import config from 'test/config';
import * as matchers from 'testUtils/matchers';
import { merge, set } from 'utils';

// we need to use require syntax to bust implicit any
const boring: FlowDefinition = require('test/flows/boring.json');

// force our matchers to be read in
const match = matchers;

export interface Resp {
  assets: { [key: string]: any }[];
}

export interface QueryString {
  [key: string]: string;
}

export const contextTypes: { [key: string]: Function } = {
  config: fakePropType,
  store: fakePropType,
  assetService: fakePropType,
};

export const baseState: AppState = mutate(initialState, {
  flowContext: merge({
    definition: require('test/flows/colors.json') as FlowDefinition,
  }),
  editorState: merge({
    language: { id: 'eng', name: 'English' },
  }),
});

const flowEditorConfig: FlowEditorConfig = config;

export const configProviderContext: ConfigProviderContext = {
  config: flowEditorConfig,
};

export const setMock = (
  implementation?: (...args: any[]) => any,
): Query<jest.Mock> => set(jest.fn(implementation));

/**
 * Compose setup method for component tests
 */
export const composeSetup = <P extends {}>(
  Component: React.ComponentClass | React.SFC,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => (
  shallowRender: boolean = true,
  propOverrides: Query<P | Partial<P>> = {},
  duxStateOverrides: Query<AppState | Partial<AppState>> = {},
  contextOverrides: Query<
    ConfigProviderContext | Partial<ConfigProviderContext>
  > = {},
  childContextTypeOverrides: { [key: string]: Function } = {},
  children: (JSX.Element | React.ComponentClass | React.SFC)[] = [],
) => {
  const props = mutate(baseProps, propOverrides);
  const store = createStore(mutate(
    baseDuxState,
    duxStateOverrides,
  ) as AppState);
  let context = mutate(baseContext, merge({ store }));

  if (Object.keys(duxStateOverrides).length > 0) {
    context = mutate(context, duxStateOverrides);
  }

  if (Object.keys(contextOverrides).length > 0) {
    context = mutate(context, contextOverrides as ConfigProviderContext);
  }

  const childContextTypes: { [contextProp: string]: Function } = mutate(
    contextTypes,
    childContextTypeOverrides,
  );
  // tslint:disable-next-line:ban-types
  const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
    <Component {...props}>{children}</Component>,
    {
      context,
      childContextTypes,
    },
  );

  return {
    wrapper,
    props,
    context,
    instance: wrapper.instance(),
  };
};

export const composeSpy = (obj: Record<string, any> | React.ComponentClass) => (
  instanceMethod: string,
) =>
  jest.spyOn(
    (obj as React.ComponentClass).prototype || obj,
    instanceMethod as any,
  );

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () =>
  new Promise((resolve: any) => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance<any>[]) =>
  spies.forEach((spy: any) => spy.mockRestore());

// To-do: type this method's output, can pass it prop generic ingested by getComponentTestUtils
/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
  componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
  specName: string,
  attributeName: string = 'data-spec',
): any => componentWrapper.find(`[${attributeName}="${specName}"]`);

export const composeDuxState = (
  query: Query<AppState | Partial<AppState>> = {},
  duxState = baseState,
) => mutate(duxState, query);

export const composeComponentTestUtils = <P extends {}>(
  Component: React.ComponentClass | React.SFC | any,
  baseProps: P = {} as any,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = configProviderContext,
) => ({
  setup: composeSetup<P>(Component, baseProps, baseDuxState, baseContext),
  spyOn: composeSpy(Component),
});

export const prepMockDuxState = (): {
  testNodes: RenderNodeMap;
  mockDuxState: AppState;
} => {
  const testNodes = getFlowComponents(boring).renderNodeMap;

  return {
    testNodes,
    mockDuxState: {
      ...initialState,
      flowContext: {
        ...initialState.flowContext,
        definition: boring,
        nodes: testNodes,
      },
    },
  };
};

export const createMockStore: Function = configureStore([thunk]);

export const mock = <T extends {}, K extends keyof T>(
  object: T,
  property: K,
  value: T[K],
) => {
  Object.defineProperty(object, property, { get: () => value });
};
