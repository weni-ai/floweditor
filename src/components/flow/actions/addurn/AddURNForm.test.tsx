import AddURNForm from 'components/flow/actions/addurn/AddURNForm';
import { Types } from 'config/interfaces';
import { AddURN } from 'flowTypes';
import React from 'react';
import { render } from 'test/utils';
import { getActionFormProps } from 'testUtils/assetCreators';

const props = getActionFormProps({
  scheme: 'tel',
  path: '+12065551212',
  type: Types.add_contact_urn,
} as AddURN);

describe(AddURNForm.name, () => {
  it('renders', () => {
    const { baseElement } = render(<AddURNForm {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
