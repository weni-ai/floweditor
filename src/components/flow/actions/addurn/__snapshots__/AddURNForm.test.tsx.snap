// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddURNForm renders 1`] = `
<body>
  <div>
    <div
      class="dialog dialog"
    >
      <div
        class="header"
      >
        <div
          class="title_container"
        >
          <div
            class="title u font secondary title-sm black"
          >
            Add URN
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <div
            class="color-neutral-cloudy"
          >
            When a contact arrives at this point in your flow...
          </div>
          <div>
            <div
              class="select_wrapper"
              data-testid="temba_select_type"
            >
              <div
                data-use-vue-component-wrap=""
                style="all: unset;"
              >
                <div
                  class="unnnic-form sm bold_placeholder"
                  data-passed-props="[object Object]"
                  data-testid="temba_select_input_type"
                  data-v-48c0c241=""
                >
                  <!---->
                  <div
                    class="unnnic-form-input text-input size--sm has-icon-right"
                    data-v-400546b3=""
                    data-v-48c0c241=""
                    mask=""
                    v-bind="[object Object]"
                  >
                    <input
                      class="input-itself input size-sm normal"
                      data-passed-props="[object Object]"
                      data-testid="temba_select_input_type"
                      data-v-400546b3=""
                      data-v-97b44e8c=""
                      iconright="arrow-button-down-1"
                      placeholder="Add a URN for the contact"
                      size="0"
                      type="text"
                      v-bind="[object Object]"
                    />
                    <!---->
                    <span
                      class="unnnic-icon unnnic-icon__size--sm unnnic-icon-scheme--neutral-cloudy icon-right"
                      data-v-400546b3=""
                      data-v-59b073b4=""
                    >
                      <svg
                        fill="none"
                        viewBox="0 0 40 40"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          class="primary"
                          d="M5.31471 14.0096L19.0432 29.5967C19.1676 29.7264 19.3169 29.8296 19.4822 29.9001C19.6474 29.9707 19.8253 30.007 20.005 30.007C20.1847 30.007 20.3625 29.9707 20.5278 29.9001C20.6931 29.8296 20.8424 29.7264 20.9668 29.5967L34.689 14.0096C34.7987 13.886 34.8825 13.7415 34.9353 13.5849C34.9882 13.4282 35.009 13.2625 34.9965 13.0977C34.984 12.9328 34.9385 12.7721 34.8627 12.6252C34.7869 12.4783 34.6824 12.3481 34.5553 12.2423L32.243 10.2963C31.9852 10.0789 31.6522 9.97163 31.316 9.99759C30.9797 10.0235 30.6671 10.1807 30.4457 10.435L20.2406 22.25C20.2112 22.2841 20.1749 22.3114 20.134 22.3302C20.0931 22.3489 20.0487 22.3587 20.0037 22.3587C19.9588 22.3587 19.9143 22.3489 19.8734 22.3302C19.8326 22.3114 19.7962 22.2841 19.7669 22.25L9.558 10.4338C9.33639 10.1806 9.02374 10.0252 8.68809 10.0013C8.52167 9.9865 8.35397 10.0052 8.19489 10.0563C8.03582 10.1074 7.8886 10.1898 7.76195 10.2988L5.44845 12.2423C5.32061 12.3479 5.21534 12.4781 5.13889 12.6252C5.06245 12.7723 5.0164 12.9332 5.0035 13.0985C4.99099 13.2635 5.01212 13.4294 5.06562 13.586C5.11911 13.7426 5.20386 13.8868 5.31471 14.0096Z"
                          fill="#3B414D"
                        />
                      </svg>
                    </span>
                  </div>
                  <!---->
                </div>
              </div>
              <div
                class="options_wrapper"
                data-testid="temba_select_options_type"
                style="margin-top: 0px; width: 0px; display: none;"
              >
                <div
                  class="option active "
                  data-option-index="0"
                >
                  <span
                    class="name"
                  >
                    Send the contact a message
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="1"
                >
                  <span
                    class="name"
                  >
                    Wait for the contact to respond
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="2"
                >
                  <span
                    class="name"
                  >
                    Smartly wait for response (New)
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="3"
                >
                  <span
                    class="name"
                  >
                    Send somebody else a message
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="4"
                >
                  <span
                    class="name"
                  >
                    Label the incoming message
                  </span>
                </div>
                <div
                  class="option  selected"
                  data-option-index="5"
                >
                  <span
                    class="name"
                  >
                    Add a URN for the contact
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="6"
                >
                  <span
                    class="name"
                  >
                    Add the contact to a group
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="7"
                >
                  <span
                    class="name"
                  >
                    Remove the contact from a group
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="8"
                >
                  <span
                    class="name"
                  >
                    Update the contact
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="9"
                >
                  <span
                    class="name"
                  >
                    Send email
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="10"
                >
                  <span
                    class="name"
                  >
                    Save a result for this flow
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="11"
                >
                  <span
                    class="name"
                  >
                    Call a webhook
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="12"
                >
                  <span
                    class="name"
                  >
                    Call Zapier
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="13"
                >
                  <span
                    class="name"
                  >
                    Enter another flow
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="14"
                >
                  <span
                    class="name"
                  >
                    Start somebody else in a flow
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="15"
                >
                  <span
                    class="name"
                  >
                    Open a ticket with a human agent
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="16"
                >
                  <span
                    class="name"
                  >
                    Send the contact airtime
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="17"
                >
                  <span
                    class="name"
                  >
                    Split by intent
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="18"
                >
                  <span
                    class="name"
                  >
                    Automatic Classify (New)
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="19"
                >
                  <span
                    class="name"
                  >
                    Split by a custom expression
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="20"
                >
                  <span
                    class="name"
                  >
                    Split by contact field
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="21"
                >
                  <span
                    class="name"
                  >
                    Split by a result in the flow
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="22"
                >
                  <span
                    class="name"
                  >
                    Split randomly
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="23"
                >
                  <span
                    class="name"
                  >
                    Split by group membership
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="24"
                >
                  <span
                    class="name"
                  >
                    Split by URN type
                  </span>
                </div>
              </div>
              <div
                class="options_wrapper"
                data-testid="temba_select_expressions_type"
                style="margin-top: 0px; width: 0px; display: none;"
              />
            </div>
          </div>
        </div>
        <div
          class="scheme_selection"
        >
          <div
            class="label u font secondary body-md color-neutral-cloudy"
          >
            Add a new URN to reach the contact
          </div>
          <div
            class="ele"
          >
            <div
              class="select_wrapper"
              data-testid="temba_select_urn_type"
            >
              <div
                data-use-vue-component-wrap=""
                style="all: unset;"
              >
                <div
                  class="unnnic-form sm bold_placeholder"
                  data-passed-props="[object Object]"
                  data-testid="temba_select_input_urn_type"
                  data-v-48c0c241=""
                  readonly="readonly"
                >
                  <!---->
                  <div
                    class="unnnic-form-input text-input size--sm has-icon-right"
                    data-v-400546b3=""
                    data-v-48c0c241=""
                    mask=""
                    v-bind="[object Object]"
                  >
                    <input
                      class="input-itself input size-sm normal"
                      data-passed-props="[object Object]"
                      data-testid="temba_select_input_urn_type"
                      data-v-400546b3=""
                      data-v-97b44e8c=""
                      iconright="arrow-button-down-1"
                      placeholder="Phone number"
                      readonly="readonly"
                      size="0"
                      type="text"
                      v-bind="[object Object]"
                    />
                    <!---->
                    <span
                      class="unnnic-icon unnnic-icon__size--sm unnnic-icon-scheme--neutral-cloudy icon-right"
                      data-v-400546b3=""
                      data-v-59b073b4=""
                    >
                      <svg
                        fill="none"
                        viewBox="0 0 40 40"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          class="primary"
                          d="M5.31471 14.0096L19.0432 29.5967C19.1676 29.7264 19.3169 29.8296 19.4822 29.9001C19.6474 29.9707 19.8253 30.007 20.005 30.007C20.1847 30.007 20.3625 29.9707 20.5278 29.9001C20.6931 29.8296 20.8424 29.7264 20.9668 29.5967L34.689 14.0096C34.7987 13.886 34.8825 13.7415 34.9353 13.5849C34.9882 13.4282 35.009 13.2625 34.9965 13.0977C34.984 12.9328 34.9385 12.7721 34.8627 12.6252C34.7869 12.4783 34.6824 12.3481 34.5553 12.2423L32.243 10.2963C31.9852 10.0789 31.6522 9.97163 31.316 9.99759C30.9797 10.0235 30.6671 10.1807 30.4457 10.435L20.2406 22.25C20.2112 22.2841 20.1749 22.3114 20.134 22.3302C20.0931 22.3489 20.0487 22.3587 20.0037 22.3587C19.9588 22.3587 19.9143 22.3489 19.8734 22.3302C19.8326 22.3114 19.7962 22.2841 19.7669 22.25L9.558 10.4338C9.33639 10.1806 9.02374 10.0252 8.68809 10.0013C8.52167 9.9865 8.35397 10.0052 8.19489 10.0563C8.03582 10.1074 7.8886 10.1898 7.76195 10.2988L5.44845 12.2423C5.32061 12.3479 5.21534 12.4781 5.13889 12.6252C5.06245 12.7723 5.0164 12.9332 5.0035 13.0985C4.99099 13.2635 5.01212 13.4294 5.06562 13.586C5.11911 13.7426 5.20386 13.8868 5.31471 14.0096Z"
                          fill="#3B414D"
                        />
                      </svg>
                    </span>
                  </div>
                  <!---->
                </div>
              </div>
              <div
                class="options_wrapper"
                data-testid="temba_select_options_urn_type"
                style="margin-top: 0px; width: 0px; display: none;"
              >
                <div
                  class="option active selected"
                  data-option-index="0"
                >
                  <span
                    class="name"
                  >
                    Phone number
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="1"
                >
                  <span
                    class="name"
                  >
                    WhatsApp Number
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="2"
                >
                  <span
                    class="name"
                  >
                    Facebook
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="3"
                >
                  <span
                    class="name"
                  >
                    Twitter ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="4"
                >
                  <span
                    class="name"
                  >
                    Telegram ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="5"
                >
                  <span
                    class="name"
                  >
                    Viber ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="6"
                >
                  <span
                    class="name"
                  >
                    Line ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="7"
                >
                  <span
                    class="name"
                  >
                    Wechat ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="8"
                >
                  <span
                    class="name"
                  >
                    Firebase ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="9"
                >
                  <span
                    class="name"
                  >
                    Jiochat ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="10"
                >
                  <span
                    class="name"
                  >
                    Freshchat ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="11"
                >
                  <span
                    class="name"
                  >
                    Email Address
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="12"
                >
                  <span
                    class="name"
                  >
                    Twitter Handle
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="13"
                >
                  <span
                    class="name"
                  >
                    External ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="14"
                >
                  <span
                    class="name"
                  >
                    Instagram ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="15"
                >
                  <span
                    class="name"
                  >
                    Slack ID
                  </span>
                </div>
                <div
                  class="option  "
                  data-option-index="16"
                >
                  <span
                    class="name"
                  >
                    Teams ID
                  </span>
                </div>
              </div>
              <div
                class="options_wrapper"
                data-testid="temba_select_expressions_urn_type"
                style="margin-top: 0px; width: 0px; display: none;"
              />
            </div>
          </div>
        </div>
        <div>
          <span
            class="label"
          >
            Value of URN
          </span>
          <div
            class="completion_wrapper"
          >
            <div
              data-testid="Value of URN"
            >
              <div
                data-use-vue-component-wrap=""
                style="all: unset;"
              >
                <div
                  class="unnnic-form sm"
                  data-passed-props="[object Object]"
                  data-v-48c0c241=""
                >
                  <!---->
                  <div
                    class="unnnic-form-input text-input size--sm"
                    data-v-400546b3=""
                    data-v-48c0c241=""
                    mask=""
                    v-bind="[object Object]"
                  >
                    <input
                      class="input-itself input size-sm normal"
                      data-passed-props="[object Object]"
                      data-v-400546b3=""
                      data-v-97b44e8c=""
                      placeholder="Enter the URN value"
                      size="12"
                      type="text"
                      v-bind="[object Object]"
                    />
                    <!---->
                    <!---->
                  </div>
                  <!---->
                </div>
              </div>
            </div>
            <div
              class="options_wrapper"
              data-testid="temba_completion_options_Value of URN"
              style="margin-top: 0px; width: 0px; display: none;"
            />
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="buttons_wrapper"
          >
            <div
              class="right_buttons"
            >
              <div
                data-use-vue-component-wrap=""
              >
                <button
                  class="unnnic-button-next unnnic-button-next--size-large unnnic-button-next--ghost unnnic-button-next-scheme--"
                  data-passed-props="[object Object]"
                  data-v-1eef067c=""
                  style="margin-left: 0px; margin-top: 0px;"
                >
                  <!---->
                  <!---->
                  <!---->
                  <span
                    class="unnnic-button-next__label"
                    data-v-1eef067c=""
                  >
                     Cancel 
                  </span>
                  <!---->
                </button>
              </div>
              <div
                data-use-vue-component-wrap=""
              >
                <button
                  class="unnnic-button-next unnnic-button-next--size-large unnnic-button-next--primary unnnic-button-next-scheme--undefined"
                  data-passed-props="[object Object]"
                  data-v-1eef067c=""
                  style="margin-top: 0px;"
                >
                  <!---->
                  <!---->
                  <!---->
                  <span
                    class="unnnic-button-next__label"
                    data-v-1eef067c=""
                  >
                     Confirm 
                  </span>
                  <!---->
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
